src/Components/Web/src/WebRenderer.cs

```diff
@@ -51,15 +76,57 @@ protected internal int AddRootComponent([DynamicallyAccessedMembers(Component)]
         /// <param name="domElementSelector">A CSS selector that uniquely identifies a DOM element.</param>
         protected abstract void AttachRootComponentToBrowser(int componentId, string domElementSelector);
 
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _interopMethodsReference.Dispose();
+            }
+
+            base.Dispose(disposing);
+        }
+
         /// <summary>
-        /// Enables support for adding, updating, and removing root components from JavaScript.
+        /// A collection of JS invokable methods that the JS-side code can use when it needs to
+        /// make calls in the context of a particular renderer. This object is never exposed to
+        /// .NET code so is only reachable via JS.
         /// </summary>
-        /// <param name="interop">The object that provides JS-callable methods for adding, updating, and removing root components.</param>
-        /// <returns>A task representing the completion of the operation.</returns>
-        protected ValueTask InitializeJSComponentSupportAsync(JSComponentInterop interop)
+        private class WebRendererInteropMethods
         {
-            var jsRuntime = _serviceProvider.GetRequiredService<IJSRuntime>();
-            return interop.InitializeAsync(jsRuntime, this);
+            private readonly WebRenderer _renderer;
+            private readonly JsonSerializerOptions _jsonOptions;
+            private readonly JSComponentInterop _jsComponentInterop;
+
+            [DynamicDependency(DynamicallyAccessedMemberTypes.PublicMethods, typeof(WebRendererInteropMethods))]
+            public WebRendererInteropMethods(WebRenderer renderer, JsonSerializerOptions jsonOptions, JSComponentInterop jsComponentInterop)
+            {
+                _renderer = renderer;
+                _jsonOptions = jsonOptions;
+                _jsComponentInterop = jsComponentInterop;
+            }
+
+            [JSInvokable]
+            public Task DispatchEventAsync(JsonElement eventDescriptor, JsonElement eventArgs)
```

Comment:
The fact that we can receive `JsonElement` as JS interop params is a cool thing I hadn't even realised would work before. It works because STJ will natively supply a `JsonElement` if you ask it to deserialize to that type.

This means that anywhere we need to do custom deserialization, we don't actually need to have manual-serialize-and-write-to-UTF8-array code on the JS side, as we can simply send normal params from JS and receive them as `JsonElement` on the .NET side. This simplifies things a lot.